### PR TITLE
Language Model Integration

### PR DESCRIPTION
**Pull request by DarshanSP19**
- Integrated [Gemini Model](https://gemini.google.com/) (Chat Generative API mode from Google) to query weather data using Natural Language Prompts.

`Input Prompt`: Daily average temperature of New York for January 2015

Relevant SQL Query:
```
SELECT 
    AVG('temperature') 
FROM 
    'gs://darshan-store/ar/2013-2022-full_37-1h-0p25deg-chunk-1.zarr-v3' 
WHERE 
    time >= '2015-01-01' AND 
    time < '2015-02-01' AND 
    city = 'New York' 
GROUP BY time_date
```
Output Data:
```
     time_date  avg_temperature
0   2015-01-01       240.978073
1   2015-01-02       243.375031
2   2015-01-03       244.584747
3   2015-01-04       249.673065
4   2015-01-05       245.650833
...
Query took: 00:01:55
```

### CODE REVIEW

File: xql/README.md [hunk 1]
```diff
@@ -113,4 +113,56 @@ _Updated on 2024-01-08_
 7. [ ] **Aliases**: Add support to alias while querying.
 8. [ ] **Join Operations**: Support joining tables and apply query.
 9. [ ] **Nested Queries**: Add support to write nested queries.
-10. [ ] **Custom Aggregate Functions**: Support custom aggregate functions
\ No newline at end of file
+10. [ ] **Custom Aggregate Functions**: Support custom aggregate functions
+
+# `weather-lm` - Querying Xarray Datasets with SQL
```
Comment (mahrsee1997) [inline xql/README.md:118]: nit: change the description 'Querying Xarray Datasets with SQL'.

File: xql/README.md [hunk 2]
```diff
@@ -113,4 +113,56 @@ _Updated on 2024-01-08_
...
+
+Querying weather data using Natural Language prompts. This uses a gemini (large language model from Google) to generate SQL like queries and `xql` to execute that query.
+> Refer to [Natural Language to Weather & Geospatial Data](https://docs.google.com/document/d/1zQ5jVIm29swVkv9nPM-vaDSatA6U3X3tv5EMATRAzU8/edit?usp=sharing&resourcekey=0-SIwltLLJD_MYkGH3H463WA) for more insights.
```
Comment (mahrsee1997) [inline xql/README.md:121]: This doc is not public, so let's remove this from here.

**Comment by DarshanSP19** [comment]
Major code changes required. Converting to draft.

File: xql/README.md (hunk 2 above)
Comment (DarshanSP19) [reply xql/README.md]: Removed.

File: xql/README.md [hunk 3]
```diff
@@ -136,4 +136,55 @@ _Updated on 2024-01-08_
 7. [ ] **Aliases**: Add support to alias while querying.
 8. [ ] **Join Operations**: Support joining tables and apply query.
 9. [ ] **Nested Queries**: Add support to write nested queries.
-10. [ ] **Custom Aggregate Functions**: Support custom aggregate functions
\ No newline at end of file
+10. [ ] **Custom Aggregate Functions**: Support custom aggregate functions
+
+# `weather-lm` - Querying Xarray Datasets with SQL
```
Comment (mahrsee1997) [inline xql/README.md:141]: nit: change the description ... its should be "Querying weather data using Natural Language prompts"

File: xql/src/weather_lm/constant.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+- "Over all locations", "globally" entails iterating through "latitude" & "longitude."
+
+Some examples of SQL queries that correspond to questions are:
+
+{few_shot_examples}
+
+Question: {question}"""
+
+few_shots = {
+    "Aggregate precipitation over months over all locations?" : "SELECT SUM(precipitation) FROM 'gs://darshan-store/ar/2013-2022-full_37-1h-0p25deg-chunk-1.zarr-v3' GROUP BY latitude, longitude, time_month",
```
Comment (mahrsee1997) [inline xql/src/weather_lm/constant.py:65]: nit: `gs://darshan-store/ar/2013-2022-full_37-1h-0p25deg-chunk-1.zarr-v3` -- can we change this to public zarr path that we have ?

File: xql/src/weather_lm/gemini.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    # Invoke pipeline to generate SQL query
+    generate_sql_res = generate_sql_model.invoke({
+        "question": input_statement,
+        "table": dataset_metadata['uri'],
+        "columns": dataset_metadata['columns'],
+        "few_shot_examples": few_shots
+    })
+
+    # Extract SQL query from result
```
Comment (mahrsee1997) [inline xql/src/weather_lm/gemini.py:68]: Can you write a comment, WHY `11:-1]`?

File: xql/src/weather_lm/gemini.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    sql_query = generate_sql_res[11:-1]
+
+    # Print generated SQL statement for debugging
+    print("SQL Statement : ", sql_query)
+
+    # Execute SQL query to fetch weather data
+    print(run_query(sql_query))
```
Comment (mahrsee1997) [inline xql/src/weather_lm/gemini.py:75]: good to have: if users were provided with the option to generate the query only.

File: xql/README.md (hunk 3 above)
Comment (DarshanSP19) [reply xql/README.md]: Done.

File: xql/src/weather_lm/constant.py (hunk 4 above)
Comment (DarshanSP19) [reply xql/src/weather_lm/constant.py]: Updated.

File: xql/src/weather_lm/gemini.py (hunk 5 above)
Comment (DarshanSP19) [reply xql/src/weather_lm/gemini.py]: Added Comment.

File: xql/src/weather_lm/gemini.py (hunk 6 above)
Comment (DarshanSP19) [reply xql/src/weather_lm/gemini.py:92]: Updated Accordingly.

Comment (mahrsee1997) [reply xql/src/weather_lm/gemini.py:92]: How to use this ?

**Review by mahrsee1997 — APPROVED** [review]
LGTM!

File: xql/src/weather_lm/gemini.py (hunk 6 above)
Comment (DarshanSP19) [reply xql/src/weather_lm/gemini.py:92]:
```
from weather_lm import nl_to_sql_query

query = nl_to_sql_query("...Natural Language Prompt...")
```

**Review by mahrsee1997 — APPROVED** [review]
LGTM!

**Pull request merged**